rocks-lib/src/upload/mod.rs

```diff
@@ -0,0 +1,177 @@
+use crate::TOOL_VERSION;
+use crate::{config::Config, project::Project};
+use itertools::Itertools as _;
+use mlua::{Lua, Value};
+use reqwest::{
+    multipart::{Form, Part},
+    Client,
+};
+use serde::Deserialize;
+use thiserror::Error;
+
+#[derive(Deserialize, Debug)]
+pub struct VersionCheckResponse {
+    version: String,
+}
+
+#[derive(Error, Debug)]
+pub enum ToolCheckError {
+    #[error(transparent)]
+    Request(#[from] reqwest::Error),
+    #[error("`rocks` is out of date with {0}'s expected tool version! `rocks` is at version {TOOL_VERSION}, server is at {}", _1.version)]
+    ToolOutdated(String, VersionCheckResponse),
+}
+
+#[derive(Error, Debug)]
+pub enum UserCheckError {
+    #[error(transparent)]
+    Request(#[from] reqwest::Error),
+    #[error("invalid API key provided")]
+    UserNotFound,
+}
+
+#[derive(Error, Debug)]
+#[error("could not check rock status on server: {0}")]
+pub struct RockCheckError(#[from] reqwest::Error);
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub enum UploadError {
+    Request(#[from] reqwest::Error),
+    #[error("unable to read rockspec: {0}")]
+    RockspecRead(#[from] std::io::Error),
+    Lua(#[from] mlua::Error),
+    ToolCheck(#[from] ToolCheckError),
+    UserCheck(#[from] UserCheckError),
+    RockCheck(#[from] RockCheckError),
+    #[error("rock already exists on server: {0}")]
+    RockExists(String),
+}
+
+pub async fn upload_from_project(
+    project: &Project,
+    api_key: String,
```

Comment:
issue: This should be a newtype with a masked `Display` implementation.

Reply:
done :)